### PR TITLE
Add Prettier check to own pipeline job

### PR DESCRIPTION
**Pull request by weshaggard**
Switch prettier check to be called by a standard devops pipeline job.

### CODE REVIEW

File: eng/scripts/Swagger-Prettier.ps1 [hunk 1]
```diff
@@ -0,0 +1,43 @@
+[CmdletBinding()]
```
Comment (konrad-jamrozik) [inline eng/scripts/Swagger-Prettier.ps1:1]: How about we rename this file to `Validate-OpenApi-Prettier`, to match `Validate-TypeSpec` ? I dislike the `Swagger Prettier` name from unified pipeline. Plus I think it will help to have different name, to avoid confusing people between old and new prettier.

File: eng/pipelines/prettier.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  pool:
+    name: azsdk-pool-mms-ubuntu-2204-general
+    vmImage: ubuntu-22.04
+
+  steps:
+  - template: /eng/pipelines/templates/steps/npm-install.yml
+
+  - pwsh: |
+      $(Build.SourcesDirectory)/eng/scripts/Swagger-Prettier.ps1 -Verbose
+    displayName: Swagger Prettier Check
```
Comment (konrad-jamrozik) [inline eng/pipelines/prettier.yml:14]: Same as I commented here:
- https://github.com/Azure/azure-rest-api-specs/pull/25718/files/71ec626a37812322300b7f15bc1ede2bde75cc53#r1320392877

Maybe `OpenAPI Prettier Valiation`

File: eng/pipelines/prettier.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+trigger: none
+
+jobs:
+- job: Prettier
```
Comment (konrad-jamrozik) [inline eng/pipelines/prettier.yml:4]: OpenAPI Prettier ?

**Comment by konrad-jamrozik** [comment]
Once we determine the check name (see [here](https://github.com/Azure/azure-rest-api-specs/pull/25718#discussion_r1320393284)) we should a corresponding entry in openapi-alps, here:
https://devdiv.visualstudio.com/DevDiv/_git/openapi-alps?path=/private/openapi-kebab/src/bots/pipeline/pipelineEventListener/checksWorkflowInfo.ts&version=GBmain&line=70&lineEnd=71&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

File: eng/scripts/Swagger-Prettier.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/scripts/Swagger-Prettier.ps1:1]: I'm open to calling this Validate-OpenApi-Prettier.

File: eng/pipelines/prettier.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/prettier.yml:14]: For the pipeline name I'm thinking we should just call it "Prettier" so that will be the name of the check but and Swagger/OpenApi is just one set of prettier checks that we may end up wanting. However if we don't think we will check any other file types then we can make this remain specific to openapi and call the pipeline "Validate OpenApi Prettier" or "OpenApi Prettier Check" seems more natural.

File: eng/pipelines/prettier.yml (hunk 3 above)
Comment (weshaggard) [reply eng/pipelines/prettier.yml:4]: We can align this with whatever we call the pipeline but currently the pipeline is where the check name comes from.

**Comment by weshaggard** [comment]
As for the name should we call this "OpenAPI Prettier Check"? @konrad-jamrozik @mikeharder what do you guys think?

File: eng/scripts/ChangedFiles-Functions.ps1 [hunk 4]
```diff
@@ -0,0 +1,56 @@
+function Get-ChangedFiles($baseCommitsh="HEAD^", $targetCommitish="HEAD", $diffFilter="d")
```
Comment (benbp) [inline eng/scripts/ChangedFiles-Functions.ps1:1]: I think the functions in here are all vulnerable to the powershell single item casting gotcha, possibly you want to [array] cast all the assigned variables in the other scripts?

**Comment by konrad-jamrozik** [comment]
> As for the name should we call this "OpenAPI Prettier Check"? @konrad-jamrozik @mikeharder what do you guys think?

Works for me! Let's agree for this to be convention and name all checks consistently, including the existing one for TypeSpec?

File: eng/scripts/ChangedFiles-Functions.ps1 (hunk 4 above)
Comment (weshaggard) [reply eng/scripts/ChangedFiles-Functions.ps1:1]: Just so I'm clear you are suggesting the callers do `@(Get-ChangedFiles)`? I believe most of the current usages should be fine but you are right new usages could get hit.

**Comment by weshaggard** [comment]
@raych1 Any ideas why my PR is failing the SDK breaking change even though the SDK generation steps didn't run and my changes aren't touching those at all.

File: eng/pipelines/prettier.yml (hunk 2 above)
Comment (mikeharder) [inline eng/pipelines/prettier.yml:14]:
```suggestion
    displayName: Swagger Prettier Check

```

File: eng/scripts/ChangedFiles-Functions.ps1 (hunk 4 above)
Comment (mikeharder) [inline eng/scripts/ChangedFiles-Functions.ps1:1]:
```suggestion
function Get-ChangedFiles($baseCommitish="HEAD^", $targetCommitish="HEAD", $diffFilter="d")
```

File: eng/scripts/ChangedFiles-Functions.ps1 [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+{
+    # diff-filter=d is to exclude any deleted files as we don't generally
```
Comment (mikeharder) [inline eng/scripts/ChangedFiles-Functions.ps1:3]: Check "TypeSpec Validation" does want to use deleted files as a signal.  If any file is deleted under folder `specification/foo`, we want to validate all TypeSpec specs under `specification/foo`, to ensure the deletion doesn't break anything.

As long as "TypeSpec Validation" can enable this by overriding the default `diff-filter`, we can make it work.  But I wonder if excluding deleted files is the correct default?

File: eng/scripts/Get-TypeSpec-Folders.ps1 [hunk 6]
```diff
@@ -1,61 +1,42 @@
...
+$checkAllPath = "specification"
+
+if ($CheckAll) {
+  $changedFiles = $checkAllPath
 }
 else {
-  Write-Host "git -c core.quotepath=off -c i18n.logoutputencoding=utf-8 diff --name-only `"$TargetBranch...$SourceBranch`" --"
-  $changedFiles = git -c core.quotepath=off -c i18n.logoutputencoding=utf-8 diff --name-only `"$TargetBranch...$SourceBranch`" --
-  $changedFiles = $changedFiles -replace '\\', '/' | Sort-Object
+  $changedFiles = Get-ChangedFiles
```
Comment (mikeharder) [inline eng/scripts/Get-TypeSpec-Folders.ps1:16]:
```suggestion
  $changedFiles = Get-ChangedFiles "HEAD^" "HEAD" ""
```

`Get-TypeSpec-Folders.ps1` needs to use deleted files as a signal to validate specs.

File: eng/pipelines/typespec-validation.yml [hunk 7]
```diff
@@ -1,32 +1,15 @@
...
-    condition: succeededOrFailed()
+  - template: /eng/pipelines/templates/steps/npm-install.yml
 
   - pwsh: |
-      $(Build.SourcesDirectory)/eng/scripts/Validate-TypeSpec.ps1 `
-        $(Build.SourcesDirectory) `
-        "origin/${env:SYSTEM_PULLREQUEST_TARGETBRANCH}" `
-        "${env:SYSTEM_PULLREQUEST_SOURCECOMMITID}" `
-        -GitClean
+      $(Build.SourcesDirectory)/eng/scripts/Validate-TypeSpec.ps1 -GitClean -Verbose
```
Comment (mikeharder) [inline eng/pipelines/typespec-validation.yml:13]: Why did you add `-Verbose` here?

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: eng/pipelines/typespec-validation.yml (hunk 7 above)
Comment (weshaggard) [reply eng/pipelines/typespec-validation.yml:13]: I had the diff helpers output more details via Write-Verbose and at least for now I wanted that output.

File: eng/scripts/ChangedFiles-Functions.ps1 (hunk 5 above)
Comment (weshaggard) [reply eng/scripts/ChangedFiles-Functions.ps1:3]: In general, for most things that are taking action on the list of files they don't want files that no longer exist because they cannot read them.

For tsv what happens in a case when an entire typspec project gets deleted? I guess you simply won't find a tspconfig file for those and it will be a noop so that should be OK for your scenario.

File: eng/pipelines/prettier.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/prettier.yml:14]: Yes that is the template that gets used they also include Stage name if there is one. I was trying to figure out how to customize so I could put all the checks in one pipeline with different jobs but I was unable to make that work.

@mikeharder so you vote would be to call this "Swagger Prettier Check" as opposed to renaming it to "OpenAPI Prettier Check"?

Comment (mikeharder) [reply eng/pipelines/prettier.yml]: I don't feel strongly about the specific check names if you want to start renaming them to be more consistent or use a new pattern.  But I think we should either keep the check names consistent for now to minimize confusion, or do a big rename at once.  I think it would be bad for spec authors and reviewers if the names are changing at random times.

**Review by mikeharder — APPROVED** [review]

**Comment by weshaggard** [comment]
I'm going to be out for the next week so I don't plan to merge this until I'm back. I agree with @mikeharder that at least during our transition we should try to keep the check names exactly the same and then after we transition we can rename them in one go. So, I will keep the check name (i.e. pipeline in this case) set to "Swagger PrettierCheck".

File: eng/pipelines/prettier.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/prettier.yml]: I agree I'll keep the "Swagger PrettierCheck" name until we finish with the larger transition.

**Comment by weshaggard** [comment]
https://dev.azure.com/azure-sdk/public/_build/results?buildId=3076795&view=results test pipeline.

**Comment by raych1** [comment]
> @raych1 Any ideas why my PR is failing the SDK breaking change even though the SDK generation steps didn't run and my changes aren't touching those at all.

@weshaggard , this is interesting. Sometimes the check would run against wrong commit which could cause this issue. But it has same result by re-running the check, so it's not the commit issue. I need to dig more for how does the `breaking change tracking` check get the result.

File: eng/scripts/ChangedFiles-Functions.ps1 [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    $changedFilesUnderSpecification = $changedFiles | Where-Object { 
+        $_.StartsWith("specification")
+    }
+    
+    return $changedFilesUnderSpecification
+}
+
+function Get-ChangedEngOrRootFiles($changedFiles = (Get-ChangedFiles))
+{
+    $engFiles = $changedFiles | Where-Object { $_.StartsWith('eng') }
```
Comment (mikeharder) [inline eng/scripts/ChangedFiles-Functions.ps1:43]:
```suggestion
    $engFiles = $changedFiles | Where-Object { $_.StartsWith('eng') }
    
    $commonTypesFiles = $changedFiles | Where-Object { $_.StartsWith('specification/common-types') }    
```

File: eng/scripts/ChangedFiles-Functions.ps1 [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    ".gitattributes",
+    ".prettierrc.json",
+    "package-lock.json",
+    "package.json",
+    "tsconfig.json"
+    )
+
+    $repoRootFiles = $changedFiles | Where-Object { $_ -in $rootFilesImpactingTypeSpec }
+
+    return $engFiles + $repoRootFiles
```
Comment (mikeharder) [inline eng/scripts/ChangedFiles-Functions.ps1:55]:
```suggestion
    return $engFiles + $commonTypesFiles + $repoRootFiles
```

File: eng/scripts/ChangedFiles-Functions.ps1 [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    # Get all the files that have changed between HEAD and the commit before head
+    # For PR's that last commit is always a merge commit so HEAD^ will get the parent
+    # commit of the base branch and as such will diff HEAD against HEAD^
+    $changedFiles = git -c core.quotepath=off diff --name-only --diff-filter=$diffFilter $baseCommitish $targetCommitish
+
+    Write-Verbose "Changed files:"
+    $changedFiles | ForEach-Object { Write-Verbose "$_" }
+
+    return $changedFiles
```
Comment (mikeharder) [inline eng/scripts/ChangedFiles-Functions.ps1:18]: Do we need to convert `\` to `/` before returning, so consumers can assume all paths use `/`?

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: eng/scripts/ChangedFiles-Functions.ps1 (hunk 8 above)
Comment (mikeharder) [reply eng/scripts/ChangedFiles-Functions.ps1:43]: Reaction to #25945

File: eng/scripts/ChangedFiles-Functions.ps1 (hunk 10 above)
Comment (weshaggard) [reply eng/scripts/ChangedFiles-Functions.ps1:18]: We don't actually need to worry about that slash now.

**Review by mikeharder — APPROVED** [review]
Changes look good, but I'd like to test this locally before you merge, if you think you are almost done.  To ensure there are no regressions in TSV scenarios.

**Comment by weshaggard** [comment]
Yes the change is ready from my testing. Once it merges I will need to enable the pipeline to automatically trigger on other PRs.

**Comment by weshaggard** [comment]
Here is the accompanying PR to remove the PrettierCheck from the unified pipeline https://dev.azure.com/devdiv/DevDiv/_git/openapi-alps/pullrequest/500709

**Comment by mikeharder** [comment]
@weshaggard: I completed my manual testing and this looks good to me.
